File: javascript/src/repositories/bucket.ts
import { z } from "zod";
import type { CallProc } from "./types.js";
import { pgBoolean, safeParse } from "./_shared.js";

export const BucketBalanceRowSchema = z
  .object({
    bucket_key: z.string().optional(),
    label: z.string().optional(),
    priority: z.number().optional(),
    expires: pgBoolean.nullable().optional(),
    balance: z
      .union([z.string(), z.number()] as const)
      .nullable()
      .optional(),
  })
  .passthrough();

export const BucketEnvelopeRowSchema = z
  .object({
    user_id: z.string().optional(),
    buckets: z.array(z.record(z.string(), z.unknown())).optional(),
    total_balance: z
      .union([z.string(), z.number()] as const)
      .nullable()
      .optional(),
  })
  .passthrough();

export const SweepRowSchema = z
  .object({
    expired_count: z.number().optional(),
    expired_amount: z
      .union([z.string(), z.number()] as const)
      .nullable()
      .optional(),
    expired_by_bucket: z
      .record(z.string(), z.union([z.string(), z.number()] as const))
      .nullable()
      .optional(),
  })
  .passthrough();

export type BucketBalanceRow = z.infer<typeof BucketBalanceRowSchema>;
export type BucketEnvelopeRow = z.infer<typeof BucketEnvelopeRowSchema>;
export type SweepRow = z.infer<typeof SweepRowSchema>;

/** Repository for credit bucket operations. */
export class BucketRepository {
  constructor(private callproc: CallProc) {}

  /** Fetch per-bucket credit balances for a user. */
  async getBucketBalances(userId: string): Promise<BucketEnvelopeRow> {
    const rows = await this.callproc("get_user_credit_buckets", [userId]);
    return safeParse(
      BucketEnvelopeRowSchema,
      rows?.[0] ?? {},
      "BucketRepository.getBucketBalances",
    );
  }

  /** Sweep expired credit grants. */
  async sweepExpiredCredits(dryRun: boolean, userId: string | null): Promise<SweepRow> {
    const rows = await this.callproc("expire_credits", [dryRun, userId]);
    return safeParse(SweepRowSchema, rows?.[0] ?? {}, "BucketRepository.sweepExpiredCredits");
  }
}

File: python/src/bursar/repositories/bucket.py
from __future__ import annotations

from bursar.repositories._types import CallProc
from bursar.repositories._utils import validate_non_empty
from bursar.repositories.schemas import BucketEnvelopeRow, SweepRow


class BucketRepository:
    """Repository for credit bucket and expiry operations.

    All methods call Postgres RPCs via the callproc function.
    Returns None when the RPC returns no rows (no result).
    Returns typed Pydantic models for successful results.
    """

    def __init__(self, callproc: CallProc) -> None:
        self._callproc = callproc

    def get_bucket_balances(self, user_id: str) -> BucketEnvelopeRow | None:
        """Get all credit bucket balances for a user.

        Args:
            user_id: The user ID.

        Returns:
            BucketEnvelopeRow if found, None if the user has no buckets.
        """
        validate_non_empty(user_id, "user_id")
        rows = self._callproc("get_user_credit_buckets", [user_id])
        if not rows:
            return None
        return BucketEnvelopeRow.model_validate(rows[0])

    def sweep_expired_credits(self, dry_run: bool, user_id: str | None) -> SweepRow | None:
        """Sweep (expire) credits that have passed their expiry date.

        Args:
            dry_run: If True, report what would be expired without modifying data.
            user_id: If set, only sweep credits for this user; otherwise sweep all.

        Returns:
            SweepRow with expiry results, or None if no credits to sweep.
        """
        rows = self._callproc("expire_credits", [dry_run, user_id])
        if not rows:
            return None
        return SweepRow.model_validate(rows[0])
